File: src/lib/hittable_list.rs
use crate::surrounding_box;
use crate::Hittable;
use crate::{Aabb, HitRecord, Ray};

#[derive(Default)]
pub struct HittableList {
    list: Vec<Box<dyn Hittable>>,
}

impl HittableList {
    #[inline(always)]
    pub fn push(&mut self, hittable: impl Hittable + 'static) {
        self.list.push(Box::new(hittable))
    }
}

impl Hittable for HittableList {
    fn hit(&self, r: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord> {
        let mut closest_so_far = t_max;
        let mut hit_anything: Option<HitRecord> = None;

        for h in self.list.iter() {
            if let Some(hit) = h.hit(r, t_min, closest_so_far) {
                closest_so_far = hit.t();
                hit_anything = Some(hit);
            }
        }

        hit_anything
    }

    fn bounding_box(&self, time0: f64, time1: f64) -> Option<Aabb> {
        let mut first_box = true;
        let mut output_box: Option<Aabb> = None;

        if !self.list.is_empty() {
            for h in self.list.iter() {
                if let Some(bounding_box) = h.bounding_box(time0, time1) {
                    output_box = if first_box {
                        Some(bounding_box)
                    } else {
                        Some(surrounding_box(output_box.unwrap(), bounding_box))
                    };
                    first_box = false;
                }
            }

            output_box
        } else {
            None
        }
    }
}

File: pkg/rtweekend/hittable_list.go
package rtweekend

type HittableList struct {
	objects []Hittable
}

func NewHittableList() *HittableList {
	hl := new(HittableList)

	hl.objects = make([]Hittable, 0)

	return hl
}

func (hl *HittableList) Clear() {
	hl.objects = []Hittable{}
}

func (hl *HittableList) Add(object Hittable) {
	hl.objects = append(hl.objects, object)
}

func (hl HittableList) Hit(r *Ray, t_min, t_max float64, rec *HitRecord) bool {
	tempRec := new(HitRecord)
	hitAnything := false
	closestSoFar := t_max

	for i := 0; i < len(hl.objects); i++ {
		if hl.objects[i].Hit(r, t_min, closestSoFar, tempRec) {
			hitAnything = true
			closestSoFar = tempRec.t
			*rec = *tempRec
		}
	}

	return hitAnything
}

func (hl HittableList) BoundingBox(time0, time1 float64, outputBox *AABB) bool {
	var tempBox AABB

	if len(hl.objects) == 0 {
		return false
	}

	firstBox := true

	for i := 0; i < len(hl.objects); i++ {
		if !hl.objects[i].BoundingBox(time0, time1, &tempBox) {
			return false
		}

		if firstBox {
			*outputBox = tempBox
		} else {
			*outputBox = *SurroundingBox(outputBox, &tempBox)
		}
		firstBox = false
	}

	return true
}
